src/Nethermind/Nethermind.Consensus/Processing/MultiCallReadOnlyBlocksProcessingEnv.cs

```diff
@@ -0,0 +1,113 @@
+
+namespace Nethermind.Consensus.Processing;
+
+public class MultiCallReadOnlyBlocksProcessingEnv : ReadOnlyTxProcessingEnvBase, IMultiCallBlocksProcessingEnv
+{
+    private readonly ITrieStore? _trieStore;
+    private readonly ILogManager? _logManager;
+    private readonly IBlockValidator _blockValidator;
+    private readonly InMemoryReceiptStorage _receiptStorage;
+    public ITransactionProcessor TransactionProcessor { get; }
+    public ISpecProvider SpecProvider { get; }
+    public MultiCallVirtualMachine Machine { get; }
+
+    //We need ability to get many instances that do not conflict in terms of editable tmp storage - thus we implement env cloning
+    public static IMultiCallBlocksProcessingEnv Create(IReadOnlyDbProvider? readOnlyDbProvider,
+        ISpecProvider? specProvider,
+        ILogManager? logManager
+        )
+    {
+        ReadOnlyDbProvider? DbProvider = new(readOnlyDbProvider, true);
+        TrieStore trieStore = new(readOnlyDbProvider.StateDb, logManager);
+        BlockTree BlockTree = new(readOnlyDbProvider,
+            new ChainLevelInfoRepository(readOnlyDbProvider.BlockInfosDb),
+            specProvider,
+            NullBloomStorage.Instance,
+            new SyncConfig(),
+            logManager);
+
+        return new MultiCallReadOnlyBlocksProcessingEnv(
+            DbProvider,
+            trieStore,
+            BlockTree,
+            specProvider,
+            logManager);
+    }
+
+    public IMultiCallBlocksProcessingEnv Clone()
+    {
+        return Create(DbProvider, SpecProvider, _logManager);
+    }
+
+    private MultiCallReadOnlyBlocksProcessingEnv(
+        IReadOnlyDbProvider? readOnlyDbProvider,
+        ITrieStore? trieStore,
+        IBlockTree? blockTree,
+        ISpecProvider? specProvider,
+        ILogManager? logManager) : base(readOnlyDbProvider, trieStore, blockTree,
+        logManager)
+    {
+
+        _trieStore = trieStore;
+        _logManager = logManager;
+        SpecProvider = specProvider;
+
+
+        _receiptStorage = new InMemoryReceiptStorage();
+
+        Machine = new MultiCallVirtualMachine(BlockhashProvider, specProvider, logManager);
+
+        HeaderValidator headerValidator = new(
+            BlockTree,
+            Always.Valid,
+            SpecProvider,
+            _logManager);
+
+        _blockValidator = new MultiCallBlockValidator(new TxValidator(SpecProvider.ChainId),
+            headerValidator,
+            Always.Valid,
+            SpecProvider,
+            _logManager);
+
+    }
+
+    public IBlockProcessor GetProcessor()
+    {
+        var yransactionProcessor = new TransactionProcessor(SpecProvider, StateProvider, Machine, _logManager);
+
+        return new BlockProcessor(SpecProvider,
+            _blockValidator,
+            NoBlockRewards.Instance,
+            new BlockProcessor.BlockValidationTransactionsExecutor(yransactionProcessor, StateProvider),
+            StateProvider,
+            _receiptStorage,
```

Comment:
Do you need to store receipts? If not just use `NullReceiptStorage`

Reply:
used